src/Lean/Server/FileWorker.lean

```diff
@@ -528,6 +528,45 @@ section RequestHandling
             children? := syms.toArray
           } :: syms', stxs'')
 
+  def handleSemanticTokens (beginPos endPos : String.Pos) :
+      ServerM (Task (Except IO.Error (Except RequestError SemanticTokens))) := do
+    let st ← read
+    let doc ← st.docRef.get
+    let text := doc.meta.text
+    let t ← doc.cmdSnaps.waitAll (·.beginPos < endPos)
+    IO.mapTask (t := t) fun (snaps, _) => do
+      let mut data := #[]
+      let mut lspPos := ⟨0, 0⟩
+      for s in snaps do
+        if s.endPos > beginPos then
+          for stx in s.stx.topDown (firstChoiceOnly := true) do
+            -- highlight keywords
+            if let Syntax.atom info val := stx then
+              if let some pos := info.getPos? then
+                if val.bsize > 0 && val[0].isAlpha && beginPos <= pos && pos < endPos then
+                  let lspPos' := text.utf8PosToLspPos pos
+                  let deltaLine := lspPos'.line - lspPos.line
+                  let deltaStart := lspPos'.character - (if lspPos'.line == lspPos.line then lspPos'.character else 0)
```

Comment:
Should maybe subtract `lspPos.character` here for same-line relative start positions?

Reply:
Good catch, this resolves the quotation and other issues mentioned above!